### PR TITLE
PWX-24702 : Add cloudOps API to upgrade OKE cluster

### PR DESCRIPTION
**Pull request by nikita-bhatia**
Jira ticket : https://portworx.atlassian.net/browse/PWX-24702

Changes done : 
> Implemented SetClusterVersion API - update kubernetes version of the Cluster
> Implemented SetInstanceGroupVersion API - update kubernetes version of nodepool and all existing nodes

Tests Done : 
> Verified Upgrade scenarion from v1.21.5 to v1.24.1 in steps and by skipping steps, for cluster , nodepool and all nodes.

### CODE REVIEW

File: oracle/oracle.go [hunk 1]
```diff
@@ -786,3 +786,92 @@ func nodePoolContainsNode(s []containerengine.Node, e string) bool {
...
+
+	if err != nil {
+		return err
+	}
+
+	err = o.waitTillWorkStatusIsSucceeded(resp.OpcRequestId, resp.OpcWorkRequestId, timeout)
+	if err != nil {
+		return err
+	}
+
```
Comment (vinayakshnd) [inline oracle/oracle.go:838]: Put a comment, why we need to scale down to 0 and scale back again. Also, add referance link to Oracle doc. that mentions the limitation.

File: oracle/oracle.go [hunk 2]
```diff
@@ -786,3 +786,92 @@ func nodePoolContainsNode(s []containerengine.Node, e string) bool {
...
+	//get existing total node count of instanceGroup
```
Comment (vinayakshnd) [inline oracle/oracle.go:839]: Put all the steps of scaling down to 0 and scaling back to original size, in separate private helper function and just call that here.

File: oracle/oracle.go [hunk 3]
```diff
@@ -786,3 +786,92 @@ func nodePoolContainsNode(s []containerengine.Node, e string) bool {
...
+		nodePoolPlacementConfigDetails[i].AvailabilityDomain = placementConfigs.AvailabilityDomain
+		nodePoolPlacementConfigDetails[i].SubnetId = placementConfigs.SubnetId
+	}
+	//delete all nodes from existing node pool
+	if err := o.SetInstanceGroupSize(instanceGroupName, 0, timeout); err != nil {
+		return err
+	}
+
+	//create same number of nodes again with updated version
+	req := containerengine.UpdateNodePoolRequest{
```
Comment (vinayakshnd) [inline oracle/oracle.go:861]: Can we use existing [SetInstanceGroupSize](https://github.com/libopenstorage/cloudops/blob/master/oracle/oracle.go#L476) function to set the size to original size?

**Review by vinayakshnd — CHANGES_REQUESTED** [review]

File: oracle/oracle.go (hunk 3 above)
Comment (nikita-bhatia) [reply oracle/oracle.go:861]: The SetInstanceGroupSize functions sets the cluster size per domain, we need functionality to create same number of instances as original. If number of instances are not in multiple of number of domains (edge case), the functionality might break. Hence not calling SetInstanceGroupSize here.

File: oracle/oracle.go (hunk 1 above)
Comment (nikita-bhatia) [reply oracle/oracle.go:838]: Added

File: oracle/oracle.go (hunk 2 above)
Comment (nikita-bhatia) [reply oracle/oracle.go:858]: Done

**Review by vinayakshnd — APPROVED** [review]
LGTM

**Pull request merged**